### ISSUE TITLE
Validate version of deploy before we start

### ISSUE THREAD

**Issue by bmorton** [bmorton/deployster#31]
Currently, you can pass any version, valid or not, to the `POST` deploys endpoint and we'll attempt to deploy that.  Should we do validation first so we can fail earlier?

If we don't do any validation, we'll get the unit deployed and it'll end up in a failed state when we go to pull it from the registry and it doesn't exist.

**Comment by ivanvanderbyl** [comment]
+1 for validation. This was my first big hiccup of using the service.

**Comment by bmorton** [comment]
Thinking out loud on ways we could do this and what complications we might have.

I haven't checked, but I'd imagine there's a API we can use for Docker hub and private registries that'd allow us to list all the tags for the image.

I was thinking authentication might be an issue, but given that we already use Docker to run tasks with deployster, we already have the Docker socket linked in and there's an expectation that authentication is setup properly for the host.  The authentication bit could probably use a little more documentation around it to make sure everything is smooth.

**Comment by bmorton** [comment]
[Here's the API](https://docs.docker.com/reference/api/registry_api/#list-repository-tags) we'd probably want to query the registry.  Not sure if this works for both public and private registries.

**Comment by ivanvanderbyl** [comment]
For private repositories we should be able to pull the auth credentials from https://docs.docker.com/reference/api/docker_remote_api_v1.17/#check-auth-configuration

**Comment by bmorton** [comment]
Do you have an example of how to pull that?  Once we have credentials, we can pull like this:

```shellsession
camisado:~$ curl -s https://registry.hub.docker.com/v1/repositories/bmorton/privaterepo/tags -H "Authorization: Basic base64creds" | jq "."
[
  {
    "layer": "6ef39a5e",
    "name": "latest"
  }
]
```

**Comment by bmorton** [comment]
Oh, and it looks like there's a client available too: https://github.com/CenturyLinkLabs/docker-reg-client

**Comment by ivanvanderbyl** [comment]
I'll have a crack at implementing this today.

**Comment by bmorton** [comment]
:+1:

**Comment by ivanvanderbyl** [comment]
So this turns out to be pretty difficult, at least for supporting private repositories, which is my main use case.

I wanted to avoid having to specify the docker hub credentials in deployster and instead load them through docker. But this doesn't seem to be possible. We could pass the `.dockercfg` to deployster in the service file, or specify those details again manually?

**Comment by bmorton** [comment]
Yeah, that's what I was afraid of.  I don't have a strong preference as to how to pass this in.

A really hacky way of doing this would be to `docker pull` the image and look at the response code?  If it fails, it'd fail fast, but if it worked, we'd be waiting for the image to download, which isn't ideal.  Probably not the right way to go, but it'd be nice if the Docker daemon surfaced this functionality somehow.

**Comment by ivanvanderbyl** [comment]
Well speaking of doing a docker pull, I thought about just using Docker's internal registry module to talk to the registry, unfortunately there's no external API for this exposed on the Docker remote API.

There would be no harm in doing a pull, accept for it blocking the request. Maybe we could return a 202, and an ID you can poll to find the status of the deploy?
